### PR TITLE
First steps towards SMS verification of profiles

### LINKED ISSUE #896: Phone verification
Using Twilio or AWS SNS (this is probably preferred due to us already using AWS heavily)

### PR DESCRIPTION
**Pull request by kalvdans**
Closes  #896

**Backend checklist**
- [x] Formatted my code by running `isort . && black .` in `app/backend`
- [x] Added tests for any new code or added a regression test if fixing a bug
- [x] All tests pass
- [ ] Run the backend locally and it works
- [x] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

- [x] check uniqueness of verified phone numbers
- [x] random Luhn token generation
- [x] hide unverified phone numbers from API

### CODE REVIEW

File: app/backend/src/couchers/errors.py
Comment (lucaslcode) [inline app/backend/src/couchers/errors.py]: These errors are displayed to the user - they won't know what E.164 is (neither do I)

File: app/pb/account.proto
Comment (lucaslcode) [inline app/pb/account.proto]: Do you mean site admins? Moderators sounds like community moderators, which would be a bad idea

File: app/pb/account.proto
Comment (lucaslcode) [inline app/pb/account.proto]:
```suggestion
  enum SMSResultCode {
```

File: app/pb/api.proto
Comment (lucaslcode) [inline app/pb/api.proto:192]: I think this value is an algorithmic value between 0 and 1, of which phone number verification is just a part?

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto]: Reading https://community.couchers.org/t/verification-process/480/18 it sounds like the phone number should be shown to the accepted hosts and accepted guests. I will change the description to that, but don't know how the API should look like yet. Probably a new API call to get all verified data.

File: app/pb/api.proto
Comment (kalvdans) [reply app/pb/api.proto:192]: Sounds good. Will adjust the comment when we implement other verification methods.

File: app/backend/src/couchers/errors.py
Comment (kalvdans) [reply app/backend/src/couchers/errors.py]: No, don't display this error to the user.

File: app/pb/account.proto
Comment (aapeliv) [inline app/pb/account.proto]: Why have these as different RPCs? I think it overly complicates the state machine, easier to just have one RPC that sets the pending number and sends a verification code, then another one to verify the sent code itself and replace the current number with the now verified one.

File: app/pb/account.proto
Comment (aapeliv) [inline app/pb/account.proto]: Error codes should be returned through gRPC StatusCodes. Also, `SMS_RESULT_CONSENT_NEEDED` is not needed, this API is "secure" so not having agreed to latest TOS means the user can never get this far.

File: app/pb/account.proto
Comment (aapeliv) [inline app/pb/account.proto:113]: Hmm, why is a checksum needed? When I get SMS codes they're these days 6 digits mostly; I think it'd be good to keep them short at the expense of not having a checksum digit.

File: app/pb/account.proto
Comment (aapeliv) [inline app/pb/account.proto]: Error codes should be handled through gRPC status codes

**Review by aapeliv — COMMENTED** [review]
I don't think the user needs to re-verify once a year, this would also incur quite a big cost to us. But I do think there should be a rate limit in how often they can update their phone number

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto]: Since you only have one try per year, I thought it would be neat to see your own phone number shown by the app before pushing the verify button. But you are right it complicates the state machine a bit.

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto]: Ok, I will spread them out evently over INVALID_ARGUMENT, FAILED_PRECONDITION, RESOURCE_EXHAUSTED and ALREADY_EXISTS.

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto:113]: I thought it would be fun to have. Frontends are free to ignore that though. 5 digits + checksum with three attempts still gives you only 0.003 % chance of guessing it.

Let's hardcode six digits in the interface.

File: app/backend/src/couchers/errors.py
Comment (lucaslcode) [reply app/backend/src/couchers/errors.py]: That would require specially handling only this error compared to the others on the frontend. 

I agree in theory, but until we switch to error codes, it will make the frontend code very messy

File: app/pb/account.proto
Comment (aapeliv) [reply app/pb/account.proto]: > Since you only have one try per year, I thought it would be neat to see your own phone number shown by the app before pushing the verify button.

We should do that in the frontend!

Comment (kalvdans) [reply app/pb/account.proto]: Yay!

**Comment by kalvdans** [comment]
Ready for review of simplified API and the implementation.

Will do migrations once you are happy with the database model!

File: app/backend/src/couchers/phone/sms.py
Comment (aapeliv) [inline app/backend/src/couchers/phone/sms.py]: I know it's a bit of a pain, but I think it'd be better to include 0s in the first digit as well (and just treat it as text), here you could zero-pad it. That way we don't loose the 10% from leading 0s.

File: app/backend/src/couchers/phone/sms.py
Comment (aapeliv) [inline app/backend/src/couchers/phone/sms.py]: I don't think there's a need to reduce the length, I think SMSs are 140 chars? Also, I've seen the first sentence be backwards so it shows up in a notification banner at the front, e.g.

"{token} is your Couchers.org verification code. If you did not requested this, please ignore this message. Best, the Couchers.org team."

Also, f-string?

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py:61]: I change this around, you'll have to rebase/merge develop here

File: app/pb/account.proto
Comment (aapeliv) [inline app/pb/account.proto:42]: Would FAILED_PRECONDITION also be appropriate here? I'm not sure

File: app/pb/account.proto
Comment (aapeliv) [inline app/pb/account.proto:51]: Maybe FAILED_PRECONDITION would be more appropriate for this as well? NOT_FOUND feels a bit funky?

I tend to use FAILED_PRECONDITION in a lot of places...

File: app/pb/account.proto
Comment (aapeliv) [inline app/pb/account.proto]: I think we should include the `+` sign; it's how people normally use phone numbers

File: app/pb/account.proto
Comment (aapeliv) [inline app/pb/account.proto:107]: I think the API should take phone numbers starting with `+`

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py]: This seems more like `is_valid_number` than `is_known_operator`?

I don't think libphonenumber checks operators, just whether it fits the known number patterns, which are different.

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py]: This should happen earlier, not after the phone has been removed from the user.

**Review by aapeliv — CHANGES_REQUESTED** [review]
API looks pretty good to me, see a few small comments!

I think we should prefix all numbers (everywhere, API and in DB) with the plus: that's what AWS expects also when we send the text, I don't see any reason to leave it out (and it also conveniently makes sure we keep them as strings everywhere).

Sorry, I ended up writing a bunch of comments! Hope they're not too pedantic and too chaotic.

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto:51]: FAILED_PRECONDITION is already taken by the "ChangePhone was not called before this method". Maybe PERMISSION_DENIED?

File: app/backend/src/couchers/servicers/account.py
Comment (kalvdans) [reply app/backend/src/couchers/servicers/account.py]: Our phonenumber database might not be up-to-date and it is a bit rude to say that my shiny new phone number is invalid.

File: app/pb/account.proto
Comment (aapeliv) [reply app/pb/account.proto:51]: The codes don't need to be unique: the frontends will just display the text anyway. There's many cases in the API where the status codes are reused but with a different message

File: app/backend/src/couchers/servicers/account.py
Comment (kalvdans) [reply app/backend/src/couchers/servicers/account.py]: Why? Clearly you have changed phone number so the earlier one is not in your possession anymore.

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto:42]: I went through the list at https://github.com/grpc/grpc/blob/master/doc/statuscodes.md and thought "The operation is not implemented or is not supported/enabled in this service." sounds just about right even if it is a third party that doesn't implement it.

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto:51]: I disagree, I'd would see FAILED_PRECONDITION as an programming error in the frontend logic here, calling methods in the wrong order.

Comment (aapeliv) [reply app/pb/account.proto:51]: I don't think there's anything that fits cleanly, INVALID_ARGUMENT implies that you supplied arguments incorrectly, FAILED_PRECONDITION indeed sounds like something is failing before the arguments are even evaluated, but I also think NOT_FOUND feels wrong too!

I think in a lot of places elsewhere I've used INVALID_ARGUMENT when the arguments are... well, invalid, e.g. if an email is not a recognized format. I've used FAILED_PRECONDITION where the arguments are correctly formatted etc but something else is wrong/missing, and NOT_FOUND when a resource was not found.

Anyway, it doesn't really matter, not a very important thing; just take your pick!

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/account.py]: Good point, I had a look at phonenumber.

Just change the error message to something that implies that the error may be fixable: "Your mobile phone number is invalid or unrecognized. Please double check it and contact support if this error persists."

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [reply app/backend/src/couchers/servicers/account.py]: Some people have multiple (active) numbers.

If the user wants to remove it because they're not in possession of it, they can do so by calling this with an empty number, we shouldn't try to "catch them" like this (even if they're not in possession of that number!).

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto:51]: I'll just stick with NOT_FOUND for now.

File: app/pb/account.proto
Comment (kalvdans) [reply app/pb/account.proto]: Let's add that such API later!

**Comment by kalvdans** [comment]
Just realized that my test file wasn't committed, but now it is included in the branch! Ready for another review and I will do migrations tomorrow unless someone does it before me :)

File: app/backend/src/couchers/models.py
Comment (aapeliv) [inline app/backend/src/couchers/models.py]: Could you move the `timedelta` into `constants.py`, and I think make it 2 years for now.

File: app/backend/src/couchers/phone/sms.py
Comment (aapeliv) [inline app/backend/src/couchers/phone/sms.py]: Please move this function into `crypto.py` (or just `"%05d" % secrets.randbelow(100000)` if you want to keep the luhn stuff here).

I want to keep all the cryptography functions in one place so they're easy to verify and audit!

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py]: Move 180 to constants too. I think we'll need to tweak these and it'll be easier to have them there

File: app/backend/src/couchers/models.py
Comment (aapeliv) [inline app/backend/src/couchers/models.py]: Could you write a brief comment here for each as to when each is null and when not and what it's used for (e.g. `phone_verification_sent` is not reset when the phone is changed, it's kept to keep track of the last time a message was sent, etc)

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py:189]: I think you also need to set `user.phone_verification_verified = None`? If the user has a phone verified before, it should be unverified now, right?

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py]: Good to also check the Luhn code here? If some frontend doesn't implement it, it's nice to not run out of attempts by writing the check digit wrong

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py]: This should be a secure compare, as insignificant as it is. (This is prone to timing attacks, though the 3 verification attempts mitigates that, still good to use a constant time compare). You can do it with `secure_compare`: that takes bytes though, you can either encode it here or add a helper function in crypto.py.

I would maybe just create `generate_sms_code()` and `verify_sms_code(request_token, right_token)` in `crypto.py`

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py]: This looks risky; are you sure it leaves the variables in the correct states? I think you should also wipe the `user.phone` and reset all other `phone_verified_*` fields as well.

**Review by aapeliv — CHANGES_REQUESTED** [review]
Thanks for adding tests! Just a few small comments left, but I think the design and API is great.

One thing that I think would be good would be to document the phone verification state machine a bit more clearly (in models.py) and then enforce that with database CHECK constraints and/or a few sanity assertions in the functions.

File: app/backend/src/couchers/servicers/account.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/account.py:194]: The could should also expire, say after 24 hours.

Comment (kalvdans) [reply app/backend/src/couchers/servicers/account.py:194]: ?SYNTAX ERROR
[edit]
You mean the verification code? It just adds complexity.

Comment (aapeliv) [reply app/backend/src/couchers/servicers/account.py:194]: Ahahahah, sorry.

I mean: the verification code should expire if the user doesn't `VerifyPhone` within some reasonable amount of time.

(That way we don't end up with this... https://bugzilla.mozilla.org/show_bug.cgi?id=1630870 [summary: GlobalSign issued a certificate with completely obsolete and non-(today's-)standards compliant cert because someone clicked a 9 year old email])

Comment (aapeliv) [reply app/backend/src/couchers/servicers/account.py:194]: (I meant this one: https://bugzilla.mozilla.org/show_bug.cgi?id=1690807 wrong cert bug lol)

File: app/backend/src/couchers/servicers/account.py
Comment (kalvdans) [reply app/backend/src/couchers/servicers/account.py:189]: Yes, thanks! Really need that state diagram :D

File: app/backend/src/couchers/models.py
Comment (kalvdans) [reply app/backend/src/couchers/models.py]: Ascii art state chart pushed! See if you can understand it and if you believe it is possible to make database constraints that checks that we are in one of the boxes. (or if you have any suggested simplification).

We could initialize times to 1970-01-01 instead of None and make them nullable=False.

Comment (aapeliv) [reply app/backend/src/couchers/models.py]: Yes, it makes sense... I need a moment to digest this :P

One thing that comes to mind is; I think `ChangePhone("")` should maybe be its own RPC (`ClearPhone`?) since it's completely decoupled?

Comment (kalvdans) [reply app/backend/src/couchers/models.py]: `ChangePhone("")` could even be removed completely. Even if I decide to be phoneless, why wouldn't I enjoy a higher verification score for 2 years for free? :D

File: app/backend/src/couchers/errors.py
Comment (kalvdans) [reply app/backend/src/couchers/errors.py]: I changed the message to "Phone number must be in international format without punctuation." But I'd expect the frontend to do some text processing before passing it to the backend so it could as well check the format itself :)

The phonenumbers Python library could do local-to-international conversion, so I assume there is one for javascript as well. :)

File: app/backend/src/couchers/models.py
Comment (aapeliv) [reply app/backend/src/couchers/models.py]: I think this is correct. Some small nitpicks:

* One missing thing is that `someone else verifies xx` should lead from any state to `Start`.
* There should maybe be some difference in the diagram between the variables in states `Code expired` and `Code sent` and similarly with `Verified` and `Verification expired`
* `Code sent` should lead to `Code sent` through `VerifyPhone(wrong)` as well

Comment (aapeliv) [reply app/backend/src/couchers/models.py]: I think it'd be good to add a db constraint to basically say `(verified is not null) => ((phone is not null) and (token is null) and (sent is not null))` and a `unique index on phone WHERE verified is not null`

Comment (aapeliv) [reply app/backend/src/couchers/models.py]: Okay, I checked the code and it seems correct. Could you add those db constraints, then I'm happy to approve this PR! (And also if you feel like tweaking the diagram, but not so important.)

Comment (kalvdans) [reply app/backend/src/couchers/models.py]: I managed the nonimportant stuff, but the constraints I failed to do. I don't think postgres has any UNIQUE .. WHERE construction.

Comment (aapeliv) [reply app/backend/src/couchers/models.py]: You can create a partial index that's also a unique index. This should work, if I got my logic right:

```py
    __table_args__ = (
        # If the phone is verified, must have a phone, and a sent time but no token
        # (verified is not null) => ((phone is not null) and (token is null) and (sent is not null))
        # note (p => q) iff (p OR ~q)
        CheckConstraint(
            "(verified IS NOT NULL) OR ((phone IS NULL) OR (token IS NOT NULL) OR (sent IS NULL))",
            name="phone_verified_conditions",
        ),
        # Verified phone numbers should be unique
        Index(
            "ix_users_unique_phone",
            phone,
            unique=True,
            postgresql_where=(verified != None),
        ),
    )
```

Comment (aapeliv) [reply app/backend/src/couchers/models.py]: Also thanks for fixing up the diagram and other comments!

Comment (kalvdans) [reply app/backend/src/couchers/models.py]: Your logic was incorrect, (p => q) iff (~p OR q), but I got the idea how to write these things. I came up with a compact expression that checks many things at once. Thanks for the secret into using an index for unique-where constraints!

File: app/backend/src/couchers/models.py
Comment (aapeliv) [inline app/backend/src/couchers/models.py:229]: I think it's cleaner to write this as you wrote in the comment:

`(phone IS NULL) OR (phone_verification_verified IS NOT NULL) OR (phone_verification_token IS NOT NULL)`

File: app/backend/src/couchers/models.py
Comment (aapeliv) [inline app/backend/src/couchers/models.py:225]: Please move `__table_args__` just after relationships, before functions for consistency, i.e. just after `blocked_user`

**Review by aapeliv — COMMENTED** [review]
Looks good! Please see the last two minor styling nitpicks and add migrations and it's good to go! Also let me know if the constraint migrations (or anything else) are giving you grief, I can also look into that.

File: app/backend/src/couchers/models.py
Comment (kalvdans) [reply app/backend/src/couchers/models.py:229]: It is not as powerful, doesn't catch (!phone && verified)

Comment (aapeliv) [reply app/backend/src/couchers/models.py:229]: Ah yeah, you're right. Sorry!

File: app/backend/src/couchers/migrations/versions/80d30951919d_phone_verifications.py
Comment (aapeliv) [inline app/backend/src/couchers/migrations/versions/80d30951919d_phone_verifications.py:68]: That's a lot of precision :P

**Review by aapeliv — DISMISSED** [review]
LGTM!

**Comment by kalvdans** [comment]
> LGTM!

... as in merge despite that I broke frontend? Just double-checking.

**Comment by aapeliv** [comment]
No, sorry, I didn't notice. I'll fix it up and push to this branch.

**Comment by kalvdans** [comment]
Thanks for a quick helping hand! Merging!

But I need another review approval, prettyplease :sad_dog_face:

**Review by aapeliv — APPROVED** [review]

File: app/backend/src/couchers/models.py
Comment (kalvdans) [inline app/backend/src/couchers/models.py:225]: This definition was overridden by a later assignment to `__table_args__`. Fixed in https://github.com/Couchers-org/couchers/pull/1428 .

File: app/backend/src/couchers/models.py
Comment (kalvdans) [inline app/backend/src/couchers/models.py:233]: Index should not be inside `__table_args__`. Fixed in https://github.com/Couchers-org/couchers/pull/1428 .

**Pull request merged**
